frontend/public/components/alert.tsx

```diff
@@ -0,0 +1,427 @@
+const alertKind = referenceForModel(AlertModel);
+const ruleKind = referenceForModel(AlertRuleModel);
+
+const detailsURL = (kind, name, labels) => `${resourcePath(kind, name)}?${_.map(labels, (v, k) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`).join('&')}`;
+
+const ExternalLink = ({href, text}) => <a className="co-external-link" href={href} target="_blank" rel="noopener noreferrer">{text}</a>;
+
+const stateToIconClassName = {
+  firing: 'fa fa-bell alert-firing',
+  pending: 'fa fa-exclamation-triangle alert-pending',
+};
+
+const State = ({state}) => {
+  if (state === 'inactive') {
+    return <span className="text-muted">{_.startCase(state)}</span>;
+  }
+  const klass = stateToIconClassName[state];
+  return klass ? <React.Fragment><i className={klass} aria-hidden="true"></i> {_.startCase(state)}</React.Fragment> : null;
+};
+
+const Annotation = ({children, title}) => _.isNil(children)
+  ? null
+  : <React.Fragment><dt>{title}</dt><dd>{children}</dd></React.Fragment>;
+
+const getURLSearchParams = () => {
+  const labels = {};
+  const params: any = new URLSearchParams(window.location.search);
+  for (let [k, v] of params.entries()) {
+    labels[k] = v;
+  }
+  return labels;
+};
+
+const connectToPrometheusURL = connectToURLs(MonitoringRoutes.Prometheus);
+
+const stateToProps = ({UI}) => UI.get(reduxID) || {};
+const connectPage = Page => withFallback(connect(stateToProps)(Page));
+
+class AlertsPageWrapper extends SafetyFirst<AlertsPageProps, any> {
+  props: AlertsPageProps;
+
+  componentDidMount () {
+    super.componentDidMount();
+
+    const {prometheusBaseURL} = (window as any).SERVER_FLAGS;
+    if (prometheusBaseURL) {
+      store.dispatch(UIActions.monitoringRulesLoading());
+
+      const poller = () => {
+        coFetchJSON(`${prometheusBaseURL}/api/v1/rules`)
+          .then(response => {
+            // Flatten the rules data to make it easier to work with and also discard non-alerting rules since those are
+            // the only ones we will be using
+            const allRules = _.flatMap(response.data.groups, 'rules');
+            const alertingRules = _.filter(allRules, {type: 'alerting'});
+            store.dispatch(UIActions.monitoringRulesLoaded(alertingRules));
+          })
+          .catch(e => store.dispatch(UIActions.monitoringRulesErrored(e)))
+          .then(() => setTimeout(() => {
+            if (this.isMounted_) {
+              poller();
+            }
+          }, 15 * 1000));
+      };
+      poller();
+    }
+  }
+}
+
+const alertStateToProps = (ignore, {match, rules}) => {
+  const name = _.get(match, 'params.name');
+  const labels = getURLSearchParams();
+
+  for (let rule of _.filter(rules, {name})) {
+    const alert = _.find(_.get(rule, 'alerts'), {labels});
+    if (alert) {
+      return {alert, name, rule};
+    }
+  }
+  return {alert: null, name, rule: _.find(rules, {name, labels})};
+};
+
+class AlertsDetailsPage_ extends AlertsPageWrapper {
```

Comment:
Skeptical of the need to use inheritance, which is usually discouraged in React because it muddies the lifecycle hooks. Prefer composition instead.